pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1130 @@
+    wanted_garbage_collected: AtomicBool,
+
+    /// Do we want to evict this layer as soon as possible? After being set to `true`, all accesses
+    /// will try to downgrade [`ResidentOrWantedEvicted`], which will eventually trigger
+    /// [`LayerInner::on_drop`].
+    wanted_evicted: AtomicBool,
+
+    /// Version is to make sure we will in fact only evict a file if no new guard has been created
+    /// for it.
+    version: AtomicUsize,
+
+    /// Allow subscribing to when the layer actually gets evicted.
+    status: tokio::sync::broadcast::Sender<Status>,
+    consecutive_failures: AtomicUsize,
+}
+
+impl std::fmt::Display for LayerInner {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.layer_desc().short_id())
+    }
+}
+
+impl AsLayerDesc for LayerInner {
+    fn layer_desc(&self) -> &PersistentLayerDesc {
+        &self.desc
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum Status {
+    Evicted,
+    Downloaded,
+}
+
+impl Drop for LayerInner {
+    fn drop(&mut self) {
+        if !*self.wanted_garbage_collected.get_mut() {
+            // should we try to evict if the last wish was for eviction?
+            // feels like there's some hazard of overcrowding near shutdown near by, but we don't
+            // run drops during shutdown (yet)
+            return;
+        }
+
+        // SEMITODO: this could be spawn_blocking or not; we are only doing the filesystem delete
+        // right now, later this will be a submit to the global deletion queue.
+        let span = tracing::info_span!(parent: None, "layer_drop", tenant_id = %self.layer_desc().tenant_id, timeline_id = %self.layer_desc().timeline_id, layer = %self);
+        let _g = span.entered();
+
+        let mut removed = false;
+        match std::fs::remove_file(&self.path) {
+            Ok(()) => {
+                removed = true;
+            }
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+                // until we no longer do detaches by removing all local files before removing the
+                // tenant from the global map, we will always get these errors even if we knew what
+                // is the latest state.
+                //
+                // we currently do not track the latest state, so we'll also end up here on evicted
+                // layers.
+            }
+            Err(e) => {
+                tracing::error!(layer = %self, "failed to remove garbage collected layer: {e}");
+            }
+        }
+
+        if let Some(timeline) = self.timeline.upgrade() {
+            if removed {
+                timeline
+                    .metrics
+                    .resident_physical_size_gauge
+                    .sub(self.layer_desc().file_size);
+            }
+            if let Some(remote_client) = timeline.remote_client.as_ref() {
+                let res =
+                    remote_client.schedule_layer_file_deletion(&[self.layer_desc().filename()]);
```

Comment:
Suspecting a deadlock in https://neon-github-public-dev.s3.amazonaws.com/reports/pr-5108/5975326246/index.html#suites/837740b64a53e769572c4ed7b7a7eeeb/48d32be3ac21dfea because of the remote timeline client usage, in case the upload task ends up being the dropper.

Easy fix to spawn_blocking this as well like it should had always been (tm), but I want to make sure I fully understand this case.